File: python/tests/test_predictive_selector.py
# SPDX-License-Identifier: MIT
"""Unit tests for PredictiveTransportSelector — Kalman RTT filter and scoring."""

from __future__ import annotations

import math
import unittest

from aether.transport.predictive_selector import PredictiveTransportSelector
from aether.transport.per_transport_metrics import PerTransportMetrics


# ── FakeTransport — minimal duck-typed stub ────────────────────────────────────

class FakeTransport:
    """Minimal duck-typed stub satisfying PredictiveTransportSelector's getattr API."""

    def __init__(
        self,
        name: str,
        bandwidth_bps: int = 500_000,
        power_cost: int = 1,
        available: bool = True,
    ) -> None:
        self.name              = name
        self.max_bandwidth_bps = bandwidth_bps
        self.power_cost_relative = power_cost
        self.is_available      = available
        self.metrics           = PerTransportMetrics()


# ── Kalman filter (indirect) ───────────────────────────────────────────────────

class TestKalmanFilterBehavior(unittest.TestCase):

    def test_kalman_converges_on_steady_state(self):
        """Feed 50 identical RTT samples — estimate should converge close to that value."""
        sel = PredictiveTransportSelector()
        t   = FakeTransport("t", available=True)
        sel.register(t, initial_rtt_ms=200.0)
        for _ in range(50):
            sel.observe_metrics(t, rtt_ms=100, success=True, bytes_transferred=1000)
        state = sel.get_kalman_state(t)
        self.assertIsNotNone(state)
        rtt_ms, _, _ = state
        self.assertAlmostEqual(rtt_ms, 100.0, delta=5.0,
                               msg=f"Kalman did not converge: rtt={rtt_ms:.2f}")

    def test_kalman_variance_decreases_with_observations(self):
        sel = PredictiveTransportSelector()
        t   = FakeTransport("t", available=True)
        sel.register(t, initial_rtt_ms=200.0)
        initial_state = sel.get_kalman_state(t)
        initial_var = initial_state[2]
        for _ in range(10):
            sel.observe_metrics(t, rtt_ms=200, success=True, bytes_transferred=1000)
        state = sel.get_kalman_state(t)
        self.assertLess(state[2], initial_var,
                        "posterior variance should decrease with observations")

    def test_kalman_detects_positive_drift(self):
        """Rising RTT should produce a positive drift estimate."""
        sel = PredictiveTransportSelector()
        t   = FakeTransport("t", available=True)
        sel.register(t, initial_rtt_ms=100.0)
        for i in range(10):
            sel.observe_metrics(t, rtt_ms=100 + (i + 1) * 15,
                                success=True, bytes_transferred=1000)
        state = sel.get_kalman_state(t)
        self.assertIsNotNone(state)
        _, drift, _ = state
        self.assertGreater(drift, 0.0,
                           f"drift {drift:.4f} should be positive for rising RTT")


# ── PredictiveTransportSelector lifecycle ─────────────────────────────────────

class TestPredictiveTransportSelector(unittest.TestCase):

    def test_register_and_rank_two_transports(self):
        sel  = PredictiveTransportSelector()
        fast = FakeTransport("fast", bandwidth_bps=1_000_000, power_cost=1,  available=True)
        slow = FakeTransport("slow", bandwidth_bps=10_000,    power_cost=10, available=True)
        sel.register(fast, initial_rtt_ms=50.0)
        sel.register(slow, initial_rtt_ms=150.0)
        # Feed good observations to fast so it has a real EWMA score.
        for _ in range(5):
            sel.observe_metrics(fast, rtt_ms=50, success=True, bytes_transferred=1000)
        ranked = sel.rank(payload_bytes=100)
        self.assertEqual(len(ranked), 2)
        self.assertEqual(ranked[0].transport.name, "fast",
                         f"expected 'fast' first, got '{ranked[0].transport.name}'")

    def test_unavailable_transport_excluded_from_ranking(self):
        sel  = PredictiveTransportSelector()
        avail  = FakeTransport("avail",  available=True)
        unavail = FakeTransport("unavail", available=False)
        sel.register(avail,   initial_rtt_ms=100.0)
        sel.register(unavail, initial_rtt_ms=100.0)
        ranked = sel.rank()
        self.assertEqual(len(ranked), 1)
        self.assertEqual(ranked[0].transport.name, "avail")

    def test_unregister_removes_transport(self):
        sel = PredictiveTransportSelector()
        t   = FakeTransport("t", available=True)
        sel.register(t, initial_rtt_ms=100.0)
        sel.unregister(t)
        self.assertEqual(len(sel.rank()), 0)

    def test_select_best_returns_none_when_empty(self):
        sel = PredictiveTransportSelector()
        self.assertIsNone(sel.select_best())

    def test_duplicate_register_ignored(self):
        sel = PredictiveTransportSelector()
        t   = FakeTransport("t", available=True)
        sel.register(t, initial_rtt_ms=100.0)
        sel.register(t, initial_rtt_ms=200.0)  # duplicate — should be no-op
        self.assertEqual(len(sel.rank()), 1)

    def test_get_kalman_state_initial_values(self):
        sel = PredictiveTransportSelector()
        t   = FakeTransport("t", available=True)
        sel.register(t, initial_rtt_ms=123.0)
        state = sel.get_kalman_state(t)
        self.assertIsNotNone(state)
        rtt, drift, variance = state
        self.assertAlmostEqual(rtt, 123.0)
        self.assertAlmostEqual(drift, 0.0)
        self.assertGreater(variance, 0.0)

    def test_get_kalman_state_unregistered_returns_none(self):
        sel = PredictiveTransportSelector()
        t   = FakeTransport("t", available=True)
        self.assertIsNone(sel.get_kalman_state(t))

    def test_scores_contain_positive_values(self):
        sel = PredictiveTransportSelector()
        t   = FakeTransport("t", available=True)
        sel.register(t, initial_rtt_ms=100.0)
        ranked = sel.rank()
        self.assertEqual(len(ranked), 1)
        self.assertGreater(ranked[0].score, 0.0)

    def test_score_increases_after_good_observations(self):
        sel = PredictiveTransportSelector()
        t   = FakeTransport("t", available=True)
        sel.register(t, initial_rtt_ms=200.0)
        score_before = sel.rank()[0].score

        # Feed 10 fast, lossless samples.
        for _ in range(10):
            sel.observe_metrics(t, rtt_ms=20, success=True, bytes_transferred=5000)

        score_after = sel.rank()[0].score
        self.assertGreater(score_after, score_before,
                           "score should improve after good observations")


if __name__ == "__main__":
    unittest.main()

File: c/tests/test_predictive_selector.c
// SPDX-License-Identifier: MIT
// Unit tests for predictive_selector.c — 2-state Kalman RTT filter and scoring.

#include <assert.h>
#include <math.h>
#include <stdbool.h>
#include <stdint.h>
#include <stdio.h>
#include <string.h>

#include "aethernet/transport.h"
#include "aethernet/predictive_selector.h"

// ── RUN macro ─────────────────────────────────────────────────────────────────

static int tests_run = 0;
#define RUN(name) do { printf("TEST: " #name "..."); name(); printf(" OK\n"); tests_run++; } while (0)

// ── Fake transport helpers ────────────────────────────────────────────────────

static bool fake_send(void *h, const char *p, const uint8_t *d, size_t n)
{
    (void)h; (void)p; (void)d; (void)n;
    return true;
}

/** Make an available transport (send != NULL). */
static void make_transport(
    aethernet_transport_vtable_t *vtable,
    aethernet_transport_t        *transport,
    const char                *name,
    int64_t                    bps,
    int32_t                    power,
    bool                       available)
{
    memset(vtable, 0, sizeof(*vtable));
    vtable->name                 = name;
    vtable->max_bandwidth_bps    = bps;
    vtable->power_cost_relative  = power;
    vtable->max_range_meters     = 100;
    vtable->send                 = available ? fake_send : NULL;
    vtable->get_metrics          = NULL;  /* rank uses default priors */

    transport->vtable = vtable;
    transport->handle = NULL;
}

// ── Kalman filter tests (via selector observe + kalman_state) ─────────────────

static void kalman_converges_on_steady_state(void)
{
    aethernet_predictive_selector_t sel;
    aethernet_predictive_selector_init(&sel);

    aethernet_transport_vtable_t vt; aethernet_transport_t t;
    make_transport(&vt, &t, "t", 500000, 1, true);
    assert(aethernet_predictive_selector_register(&sel, &t, 200.0));

    for (int i = 0; i < 50; i++)
        aethernet_predictive_selector_observe(&sel, &t, 100, true, 1000);

    double rtt, drift, variance;
    assert(aethernet_predictive_selector_kalman_state(&sel, &t, &rtt, &drift, &variance));
    /* After 50 identical observations the estimate must be within 5 ms of 100. */
    assert(fabs(rtt - 100.0) < 5.0);
}

static void kalman_variance_decreases_with_observations(void)
{
    aethernet_predictive_selector_t sel;
    aethernet_predictive_selector_init(&sel);

    aethernet_transport_vtable_t vt; aethernet_transport_t t;
    make_transport(&vt, &t, "t", 500000, 1, true);
    assert(aethernet_predictive_selector_register(&sel, &t, 200.0));

    /* Record initial variance (before any observations). */
    double rtt0, drift0, var0;
    assert(aethernet_predictive_selector_kalman_state(&sel, &t, &rtt0, &drift0, &var0));

    for (int i = 0; i < 10; i++)
        aethernet_predictive_selector_observe(&sel, &t, 200, true, 1000);

    double rtt1, drift1, var1;
    assert(aethernet_predictive_selector_kalman_state(&sel, &t, &rtt1, &drift1, &var1));
    /* Posterior variance must be strictly smaller than the prior. */
    assert(var1 < var0);
}

static void kalman_detects_positive_drift_for_rising_rtt(void)
{
    aethernet_predictive_selector_t sel;
    aethernet_predictive_selector_init(&sel);

    aethernet_transport_vtable_t vt; aethernet_transport_t t;
    make_transport(&vt, &t, "t", 500000, 1, true);
    assert(aethernet_predictive_selector_register(&sel, &t, 100.0));

    for (int i = 0; i < 10; i++)
        aethernet_predictive_selector_observe(&sel, &t, (uint64_t)(100 + (i + 1) * 15), true, 1000);

    double rtt, drift, variance;
    assert(aethernet_predictive_selector_kalman_state(&sel, &t, &rtt, &drift, &variance));
    /* Rising RTT must produce positive drift estimate. */
    assert(drift > 0.0);
}

// ── Selector lifecycle tests ──────────────────────────────────────────────────

static void register_and_rank_fast_transport_first(void)
{
    aethernet_predictive_selector_t sel;
    aethernet_predictive_selector_init(&sel);

    aethernet_transport_vtable_t vt_fast, vt_slow;
    aethernet_transport_t t_fast, t_slow;
    make_transport(&vt_fast, &t_fast, "fast", 1000000L, 1, true);
    make_transport(&vt_slow, &t_slow, "slow", 10000L,   10, true);

    assert(aethernet_predictive_selector_register(&sel, &t_fast, 50.0));
    assert(aethernet_predictive_selector_register(&sel, &t_slow, 150.0));

    /* Feed a few good observations to the fast transport to confirm ranking. */
    for (int i = 0; i < 5; i++)
        aethernet_predictive_selector_observe(&sel, &t_fast, 50, true, 1000);

    aethernet_predictive_rank_entry_t ranked[AETHERNET_PREDICTIVE_MAX_TRANSPORTS];
    size_t count = 0;
    aethernet_predictive_selector_rank(&sel, 100, ranked, &count);

    assert(count == 2);
    assert(ranked[0].transport == &t_fast);
    assert(ranked[0].score > 0.0);
}

static void unavailable_transport_excluded_from_rank(void)
{
    aethernet_predictive_selector_t sel;
    aethernet_predictive_selector_init(&sel);

    aethernet_transport_vtable_t vt_avail, vt_unavail;
    aethernet_transport_t t_avail, t_unavail;
    make_transport(&vt_avail,   &t_avail,   "avail",   500000L, 1, true);
    make_transport(&vt_unavail, &t_unavail, "unavail", 500000L, 1, false);

    assert(aethernet_predictive_selector_register(&sel, &t_avail,   100.0));
    assert(aethernet_predictive_selector_register(&sel, &t_unavail, 100.0));

    aethernet_predictive_rank_entry_t ranked[AETHERNET_PREDICTIVE_MAX_TRANSPORTS];
    size_t count = 0;
    aethernet_predictive_selector_rank(&sel, 0, ranked, &count);

    assert(count == 1);
    assert(ranked[0].transport == &t_avail);
}

static void unregister_removes_transport(void)
{
    aethernet_predictive_selector_t sel;
    aethernet_predictive_selector_init(&sel);

    aethernet_transport_vtable_t vt; aethernet_transport_t t;
    make_transport(&vt, &t, "t", 500000, 1, true);
    assert(aethernet_predictive_selector_register(&sel, &t, 100.0));

    aethernet_predictive_selector_unregister(&sel, &t);

    aethernet_predictive_rank_entry_t ranked[AETHERNET_PREDICTIVE_MAX_TRANSPORTS];
    size_t count = 0;
    aethernet_predictive_selector_rank(&sel, 0, ranked, &count);
    assert(count == 0);
}

static void select_best_returns_null_when_empty(void)
{
    aethernet_predictive_selector_t sel;
    aethernet_predictive_selector_init(&sel);
    assert(aethernet_predictive_selector_best(&sel, 100) == NULL);
}

static void duplicate_register_is_noop(void)
{
    aethernet_predictive_selector_t sel;
    aethernet_predictive_selector_init(&sel);

    aethernet_transport_vtable_t vt; aethernet_transport_t t;
    make_transport(&vt, &t, "t", 500000, 1, true);
    assert(aethernet_predictive_selector_register(&sel, &t, 100.0));
    assert(!aethernet_predictive_selector_register(&sel, &t, 200.0)); /* duplicate: false */

    aethernet_predictive_rank_entry_t ranked[AETHERNET_PREDICTIVE_MAX_TRANSPORTS];
    size_t count = 0;
    aethernet_predictive_selector_rank(&sel, 0, ranked, &count);
    assert(count == 1);
}

static void kalman_state_initial_values_are_correct(void)
{
    aethernet_predictive_selector_t sel;
    aethernet_predictive_selector_init(&sel);

    aethernet_transport_vtable_t vt; aethernet_transport_t t;
    make_transport(&vt, &t, "t", 500000, 1, true);
    assert(aethernet_predictive_selector_register(&sel, &t, 123.0));

    double rtt, drift, variance;
    assert(aethernet_predictive_selector_kalman_state(&sel, &t, &rtt, &drift, &variance));
    assert(fabs(rtt   - 123.0) < 1e-9);
    assert(fabs(drift - 0.0)   < 1e-9);
    assert(variance > 0.0);
}

static void kalman_state_unregistered_transport_returns_false(void)
{
    aethernet_predictive_selector_t sel;
    aethernet_predictive_selector_init(&sel);

    aethernet_transport_vtable_t vt; aethernet_transport_t t;
    make_transport(&vt, &t, "t", 500000, 1, true);
    /* Never register — state query must return false. */

    double rtt, drift, variance;
    assert(!aethernet_predictive_selector_kalman_state(&sel, &t, &rtt, &drift, &variance));
}

static void rank_returns_positive_score(void)
{
    aethernet_predictive_selector_t sel;
    aethernet_predictive_selector_init(&sel);

    aethernet_transport_vtable_t vt; aethernet_transport_t t;
    make_transport(&vt, &t, "t", 500000, 1, true);
    assert(aethernet_predictive_selector_register(&sel, &t, 100.0));

    aethernet_predictive_rank_entry_t ranked[AETHERNET_PREDICTIVE_MAX_TRANSPORTS];
    size_t count = 0;
    aethernet_predictive_selector_rank(&sel, 0, ranked, &count);
    assert(count == 1);
    assert(ranked[0].score > 0.0);
}

static void score_improves_after_good_observations(void)
{
    aethernet_predictive_selector_t sel;
    aethernet_predictive_selector_init(&sel);

    aethernet_transport_vtable_t vt; aethernet_transport_t t;
    make_transport(&vt, &t, "t", 500000, 1, true);
    assert(aethernet_predictive_selector_register(&sel, &t, 200.0));

    aethernet_predictive_rank_entry_t ranked[AETHERNET_PREDICTIVE_MAX_TRANSPORTS];
    size_t count = 0;
    aethernet_predictive_selector_rank(&sel, 0, ranked, &count);
    assert(count == 1);
    double score_before = ranked[0].score;

    for (int i = 0; i < 10; i++)
        aethernet_predictive_selector_observe(&sel, &t, 20, true, 5000);

    aethernet_predictive_selector_rank(&sel, 0, ranked, &count);
    assert(count == 1);
    double score_after = ranked[0].score;

    assert(score_after > score_before);
}

// ── main ──────────────────────────────────────────────────────────────────────

int main(void)
{
    printf("Aether Predictive Selector — Unit Tests\n");
    printf("========================================\n");

    RUN(kalman_converges_on_steady_state);
    RUN(kalman_variance_decreases_with_observations);
    RUN(kalman_detects_positive_drift_for_rising_rtt);
    RUN(register_and_rank_fast_transport_first);
    RUN(unavailable_transport_excluded_from_rank);
    RUN(unregister_removes_transport);
    RUN(select_best_returns_null_when_empty);
    RUN(duplicate_register_is_noop);
    RUN(kalman_state_initial_values_are_correct);
    RUN(kalman_state_unregistered_transport_returns_false);
    RUN(rank_returns_positive_score);
    RUN(score_improves_after_good_observations);

    printf("\n%d tests passed.\n", tests_run);
    return 0;
}
